### PR TITLE
Reynolds stress

### PR DESCRIPTION
**Pull request by acdaigneault**
This branch adds a Reynolds stress calculation. We can display Reynolds stresses (<u'u'>, <v'v'>, <w'w'>,<u'v'>,<v'w'>,<w'u'>)  in Paraview like other solutions.
It works for 2d/3d and for both gls_navier_stokes and gd_navier_stokes solvers.
There's 2 unit tests, the first one tests 3d with Trilinos vectors and the second one tests 2d with Trilinos block vectors with adaptive time steps.

### CODE REVIEW

File: include/solvers/postprocessing_velocities.h
Comment (blaisb) [inline include/solvers/postprocessing_velocities.h:123]:
```suggestion
  VectorType sum_reynolds_stress_dt; +x
```

File: examples/07-periodic_hills/periodic_hills.prm
Comment (blaisb) [inline examples/07-periodic_hills/periodic_hills.prm:9]: No need for such a complex output name. You can just call it periodic_hill_output  or just hill_output or whatever. Make it simple :)

File: examples/07-periodic_hills/periodic_hills.prm
Comment (blaisb) [inline examples/07-periodic_hills/periodic_hills.prm:20]: Yay!

File: examples/07-periodic_hills/periodic_hills.prm
Comment (blaisb) [inline examples/07-periodic_hills/periodic_hills.prm:57]: This would be how it would be done in the real case? For the tutorial it would be better to have it done sooner, even if this will not be the "real" way it is done. The idea  is still to be able to run the tutorial for a limited amount of time like 30-minutes or 1 hour.

File: include/core/parameters.h
Comment (blaisb) [inline include/core/parameters.h:243]: I wonder if we should really add this. I mean, when you calculate averages you will most likely calculate_average_velocities we will also want the Reynolds stresses. I'm not sure we need an additional parameter for this shenanigan :)

File: source/solvers/postprocessing_velocities.cc
Comment (blaisb) [inline source/solvers/postprocessing_velocities.cc:82]: Love this

File: source/solvers/postprocessing_velocities.cc
Comment (blaisb) [inline source/solvers/postprocessing_velocities.cc:89]: Ohh wow. Popping up the fancy constexpr? This is the good stuff.

File: source/solvers/postprocessing_velocities.cc
Comment (blaisb) [inline source/solvers/postprocessing_velocities.cc:96]: Why loop with increments of one then? You could loop by increments of dim+1 : " i+= dim+1". C++ allows thats. You can do whatever you want to i.

File: source/solvers/postprocessing_velocities.cc
Comment (blaisb) [inline source/solvers/postprocessing_velocities.cc:155]: Great, this is legit.

File: source/solvers/postprocessing_velocities.cc
Comment (blaisb) [inline source/solvers/postprocessing_velocities.cc:15]: The idea here I guess is to only calculate average when you really need it right? Why not calculate it then when we ask for it instead? (in the getter) this would prevent this argument being passed here. To me it is weird to have a calculation routine depending on an i/o argument.

File: source/solvers/postprocessing_velocities.cc
Comment (blaisb) [inline source/solvers/postprocessing_velocities.cc:44]: RIght now this means that you average velocity calculation won't work with dynamic mesh adaptation. You should at least document that because when we refine the mesh we need to "reinit the vectors and transfer the solution". It's not mandatory right now, but clearly this support should be added eventually.

**Review by blaisb — APPROVED** [review]
Looks very good to me. Very good work.
Things to do:

- Add documentation to the Wiki
- Verify that you really need to pass is_output_iteration every calculation or not (why not calculate it only when you need a getter). Additionnaly, don't you always need to have the average velocity freshen up for reynold stresses?
- No need for two bools, just calculate both Reynolds stresses and average, no need for two bools I think
- Clearly document that the routine will fail if mesh adaptation is used. This is something we can fix in a next PR by adding a "adapt to new mesh" routine to the averaging class. We can look at this together
- Finally, this is very very good work!

File: examples/07-periodic_hills/periodic_hills.prm
Comment (acdaigneault) [reply examples/07-periodic_hills/periodic_hills.prm:57]: I should be more careful about committing the final .prm in my last push, my bad!

File: source/solvers/postprocessing_velocities.cc
Comment (acdaigneault) [reply source/solvers/postprocessing_velocities.cc:44]: Do you think I should add a condition where if there's a dynamic mesh adaptation all vectors reinitializes at each iteration?

File: source/solvers/postprocessing_velocities.cc
Comment (acdaigneault) [reply source/solvers/postprocessing_velocities.cc:15]: My idea was to only calculate the average_velocities = sum__velocities_dt / dt when we want the data out or if reynolds stress is enable (because it needs the average velocities at each iteration). If calculate_reynolds_stress is removed, average_velocities should be calculated everytime so no needs to have an boolean of something in the getter to calculate it anymore.

File: source/solvers/postprocessing_velocities.cc
Comment (acdaigneault) [reply source/solvers/postprocessing_velocities.cc:96]: Because if begin_index is not 0, dim+1, 2(dim+1), ... i will never get the right value for the vector indexation. I know it's kinda weird, it's because of the <u'v'> calculation. I tried it and it didn't work, but I will try again to make sure.

Comment (acdaigneault) [reply source/solvers/postprocessing_velocities.cc:96]: I did try and it did not work, but I must have done something wrong because it actually works mhm.

File: source/solvers/navier_stokes_base.cc
Comment (blaisb) [inline source/solvers/navier_stokes_base.cc:895]: This is not correct. You want to assert throw nsparam.mesh_adaptation.type == none.
What this means is AssertThrow(nsparam.mesh_adaptation.type==none,...) If the condition is not true, it will throw and error and stop execution of the program. Right now this is working because you are throwing true, but true will always be true and thus always pass.

**Review by blaisb — APPROVED** [review]
Just a tiny mistake with the AssertThrow, but I will fix it before merging :) !
Great work! Looking forward to launching this on Niagara :)! Can't wait

File: source/solvers/navier_stokes_base.cc
Comment (acdaigneault) [reply source/solvers/navier_stokes_base.cc:895]: Ohhhhhh yeah I get it!

**Comment by acdaigneault** [comment]
> Just a tiny mistake with the AssertThrow, but I will fix it before merging :) !
> Great work! Looking forward to launching this on Niagara :)! Can't wait

Me too :) !!

**Comment by blaisb** [comment]
Some elements still need to be fixed :

1. Reynold stresses should be interpreted as tensors and not as vector + scalaer
2. Two components of the Reynolds stresses are missing. The full tensor should be calculated.
3. There are some places where you pass your vectors by copies instead of by reference. For example 
```
  const VectorType
  get_average_velocities()
  {
    return average_velocities;
  }

  /**
   * @brief get_reynolds_stress. Gives the time-averaged normal Reynolds
   * stresses and shear stress
   */
  const VectorType
  get_reynolds_stress()
  {
    return reynolds_stress;
  }
```

**Comment by acdaigneault** [comment]
Last commits implement the Reynolds stresses independent components instead of only normal Reynolds stresses and shear stress.
There's new objets related to it as dof_hanfler_rs and fe_rs. There are needed to allow data out because que number of independent components is not the same as the number of dof per node.

**Review by blaisb — APPROVED** [review]
Ahh, like a melody to my eyes.
Good work. I will merge it tomorrow once Actions shenanigans are finished with.

**Pull request merged**
